compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2384,7 +2385,23 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                         Some((span, open_paren_sp)) => Some((span, true, open_paren_sp)),
                         // If `param` corresponds to `Self`, no usable suggestion span.
                         None if generics.has_self && param.index == 0 => None,
-                        None => Some((self.tcx.def_span(def_id).shrink_to_hi(), false, None)),
+                        None => {
+                            let span = self.tcx.def_span(def_id);
+                            let span = if type_param.default_value(self.tcx).is_some() {
+                                let start = span.shrink_to_lo();
+                                self.tcx
+                                    .sess
+                                    .source_map()
+                                    .span_extend_while(start, |char| {
+                                        !char.is_whitespace() && char != '='
+                                    })
+                                    .unwrap()
+                                    .shrink_to_hi()
+                            } else {
+                                span.shrink_to_hi()
+                            };
```

Comment:
The results are great, but could we instead of using textual checks take advantage that we have the HIR node for the parameters already here?

Reply:
Actually, the value of `﻿hir_generics.predicates` is an empty array, and we do not store the ﻿span of `﻿=` for these cases :(

> that way ensures...

Yes, I agree. Using sourcemap can lead to potential issues when dealing with macros.

Reply:
That's surprising, can you tell me the contents of `hir_generics` here?

Reply:
I used ﻿`tracing` to print the value:
- The `﻿hir_generics` for `﻿T = i32` in `﻿struct Foo<T = i32>(&'static T)` is:[log](https://github.com/rust-lang/rust/files/15401686/log2.log)
- By comparison, the `﻿hir_generics` for `﻿T: IntoIterator = Vec<i32>` in ﻿`struct Foo<T: IntoIterator = Vec<i32>>(&'static T)` is [here](https://github.com/rust-lang/rust/files/15401707/log3.log), which will eventually trigger this [branch](https://github.com/rust-lang/rust/pull/124884/files#diff-845d26194de711d60e1317dbe90812d5d7635f4f758222fd7e14bf550ffe895dR2385).

Reply:
Ah, I see. It seems like what you have to do is look for the appropriate param in `hir_generics.params`, and see if `if let Type { default: Some(ty), .. } = param.kind`, and use that `ty`'s span.